### PR TITLE
fix: use singleton mlir context

### PR DESCRIPTION
**Pull request by polvalente**
With a little bit of back and forth with @jpienaar from OpenXLA,
turns out that MLIRContext manages threading, so using a global singleton
is better for performance.

MLIR is now on perf. parity with XlaOp from my local benchmarks.

This also explains why we saw no significant difference when running a problematic test by itself.

### CODE REVIEW

File: exla/test/test_helper.exs [hunk 1]
```diff
@@ -26,6 +26,8 @@ if client.platform == :host and client.device_count == 1 and System.schedulers_o
   )
 end
 
+IO.gets("Press enter to continue... PID: #{System.pid()}}")
+
```
Comment (polvalente) [inline exla/test/test_helper.exs:30]:
```suggestion
```

File: exla/lib/exla/mlir/module.ex [hunk 2]
```diff
@@ -11,11 +11,22 @@ defmodule EXLA.MLIR.Module do
...
   @doc """
   Creates a new MLIR module.
   """
   def new() do
-    ref = EXLA.NIF.new_mlir_module() |> unwrap!()
+    context = :persistent_term.get(@context_key, nil)
+
+    unless context do
+      raise "No MLIR context found"
+    end
```
Comment (jonatanklosko) [inline exla/lib/exla/mlir/module.ex:27]: We could do the init lazily:

```elixir
defp get_or_create_context() do
  if context = :persistent_term.get(@context_key, nil) do
    context
  else
    context = EXLA.NIF.new_mlir_context() |> unwrap!()
    :persistent_term.put(@context_key, context)
    context
  end
end
```

In the very unlikely race condition, where `new/0` is called concurrently we may create two contexts, which should be fine.

@josevalim?

Comment (polvalente) [reply exla/lib/exla/mlir/module.ex:27]: I changed exactly because there was a race condition

Comment (jonatanklosko) [reply exla/lib/exla/mlir/module.ex:27]: Ok, nevermind then :D

File: exla/lib/exla/mlir/module.ex [hunk 3]
```diff
@@ -11,11 +11,22 @@ defmodule EXLA.MLIR.Module do
   alias EXLA.Executable
   alias EXLA.Shape
 
+  @context_key {__MODULE__, :mlir_context}
+  def init do
+    :persistent_term.put(@context_key, EXLA.NIF.new_mlir_context() |> unwrap!())
+  end
```
Comment (josevalim) [inline exla/lib/exla/mlir/module.ex:17]: What if we initialize one context per client and then we shove this in the client?

Comment (polvalente) [reply exla/lib/exla/mlir/module.ex:17]: That might work. I'll try it out

Comment (polvalente) [reply exla/lib/exla/mlir/module.ex:17]: I tried initializing this inside the client at the C++ level and that led to some race conditions which ended in segfaults.

Seeing as different contexts will eat up more threads, I'm not sure there's any gain here outside of the code seeming more isolated.

Comment (polvalente) [reply exla/lib/exla/mlir/module.ex:17]: Actually, I just realized that multi-device is broken in this implementation already, so I'll have to figure something else out

File: exla/c_src/exla/mlir/builder.cc [hunk 4]
```diff
@@ -1315,6 +1309,9 @@ void MLIRModule::LowerPatterns() {
   mlir::stablehlo::populateStablehloToHloPatterns(&patterns, &converter, context());
 
   mlir::applyPartialConversion(module(), target, std::move(patterns));
+
+  target.addLegalDialect<mlir::stablehlo::StablehloDialect>();
+  target.addLegalDialect<mlir::func::FuncDialect>();
```
Comment (polvalente) [inline exla/c_src/exla/mlir/builder.cc:1314]:
```suggestion
```

File: exla/lib/exla/mlir/module.ex [hunk 5]
```diff
@@ -14,23 +15,35 @@ defmodule EXLA.MLIR.Module do
...
-      )
-      when is_binary(name) do
+  def add_function(module, name, arg_shapes, return_shapes) do
+    create_function(module, name, arg_shapes, return_shapes, false)
+  end
+
+  # @doc """
+  # Creates a new MLIR function with the given name belonging
+  # to the given MLIR module.
+  # """
```
Comment (josevalim) [inline exla/lib/exla/mlir/module.ex:38]:
```suggestion
```

File: exla/lib/exla/mlir/context_pool.ex [hunk 6]
```diff
@@ -0,0 +1,37 @@
+defmodule EXLA.MLIR.ContextPool do
+  @doc false
```
Comment (josevalim) [inline exla/lib/exla/mlir/context_pool.ex:2]:
```suggestion
  @moduledoc false
```

Comment (josevalim) [reply exla/lib/exla/mlir/context_pool.ex:2]: Btw, let's `@moduledoc false` all MLIR modules, cause their APIs will certainly change once we remove xla mode.

**Review by josevalim — APPROVED** [review]

**Pull request merged**